src/main/java/com/google/graphgeckos/dashboard/storage/DataRepository.java

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.graphgeckos.dashboard.storage;
+
+/**
+ * An interface for providing unified I/O to the storage component.
+ */
+public interface DataRepository {
+  /**
+   * Creates a new database entry, with just the metadata of a commit.
+   *
+   * @param entryData a JSON in plain string format, must have a "commitHash" field.
+   * @throws IllegalArgumentException if entryData is null or doesn't have a
+   *      "commitHash" field.
+   */
+  void createNewEntry(final ParsedGitData entryData);
+
+  /**
+   * Updates an existing database entry, with the individual information from
+   * a particular buildbot.
+   *
+   * @param updateData a JSON in plain string format, must have a "commitHash" field.
+   * @throws IllegalArgumentException if entryData is null or doesn't have a
+   *     "commitHash" field.
+   */
+  void updateExistingEntry(final ParsedBuildbotData updateData);
+
+  /**
+   * Deletes a database entry which has its commit hash equal to the one provided in the
+   * arguments.
+   *
+   * @param commitHash the String representation of the commit hash of the revision data to
+   * be deleted.
+   */
+  void deleteEntryByCommitHash(final String commitHash);
+
+  /**
+   * Queries the database for a specified amout of entries, going down in chronological
+   * order, starting from an offest.
+   *
+   * @param number is the number of database entries to retrieve.
+   * @param offset is the offset from the latest database entry, for which to consider
+   *     the requested number of entries.
+   * @return an iterable object containing at most {@code number} entries starting
+   *     from the latest entry - {@code offset}. If the database has not enough entries
+   *     for the requested {@code offset} and {@code number}, returns all the available
+   *     entries from that range.
+   * @throws IllegalArgumentException if either number or offset are < 0.
+   */
+  Iterable<AggregatedBuildbotData> getLastEntries(final int number, final int offset)
+                                                         throws IllegalArgumentException {
```

Comment:
I believe `IllegalArgumentException` is not appropriate as a checked exception.
Conventionally, subclasses of `RuntimeException` are unchecked.
`IllegalArgumentException` is a subclass of `RuntimeException`: http://developer.classpath.org/doc/java/lang/IllegalArgumentException-source.html
https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

If you need to have this be checked, some direct subclass of `Exception`, possibly user-defined, would be more appropriate.

That being said, throwing an `IllegalArgumentException` makes more sense to me when `number < 0` or `offset < 0`. Ensuring these invariants is fully within the powers of the caller code and I would consider calling this with negative args just a bug in the calling code. Note that declaring this checked means that at each point calling code calls this, it will have to explicitly handle the IllegalArgumentException, either by a try-catch block or by trowing itself. I think a more clean approach is for this to be unchecked and leave calling code establish the invariants for the arguments before calling this.
For example, suppose we call this form a place that takes arguments from user data. I'd prefer (1) over (2).
```
// 1
public Json getUsages(int userSuppliedNumber, int userSuppliedOffset) {
  if (userSuppliedNumber < 0 || userSuppliedOffest < 0) {
    // do some default behavior, such as returning empty response
    ...;
    return ...;
  }
  // this is safe
  Iterable<> entries = storage.getLastEntries(userSuppliedNumber, userSuppliedOffest);
  // do some processing
}

// 2
public Json getUsages(int userSuppliedNumber, int userSuppliedOffset) {
  try {
    // this is unsafe
    Iterable<> entries = storage.getLastEntries(userSuppliedNumber, userSuppliedOffest);
    // do some processing
  } catch (SomeException e) {
    // do some default behavior
  }
}
```

Reply:
Declaring a method as `throws IllegalArgumentException` doesn't make this a checked exception. It's essentially only (good) documentation and appropriate.

Reply:
Thank you Michael!

I was wrong in assuming the throws declaration make it a checked exception. The way it's declared now is good!

Reply:
Having said that, you should be consistent about adding a `throws` clause or not.